lib/contextual_logger.rb

```diff
@@ -5,94 +5,98 @@
   end
 
-  def warn(progname = nil, **extra_context, &block)
-    add(Logger::Severity::WARN, nil, progname, extra_context, &block)
+  def error(message = nil, extra_context = {})
+    add_if_enabled(Logger::Severity::ERROR, message || yield, extra_context: extra_context)
   end
 
-  def error(progname = nil, **extra_context, &block)
-    add(Logger::Severity::ERROR, nil, progname, extra_context, &block)
+  def fatal(message = nil, extra_context = {})
+    add_if_enabled(Logger::Severity::FATAL, message || yield, extra_context: extra_context)
   end
 
-  def fatal(progname = nil, **extra_context, &block)
-    add(Logger::Severity::FATAL, nil, progname, extra_context, &block)
+  def unknown(message = nil, extra_context = {})
+    add_if_enabled(Logger::Severity::UNKNOWN, message || yield, extra_context: extra_context)
   end
 
-  def unknown(progname = nil, **extra_context, &block)
-    add(Logger::Severity::UNKNOWN, nil, progname, extra_context, &block)
+  def log_level_enabled?(severity)
+    severity >= @level
   end
 
-  def add(severity, message = nil, progname = nil, extra_context = nil)
-    severity ||= UNKNOWN
-    if @logdev.nil? || (severity < @level)
-      return true
+  def add_if_enabled(severity, message, extra_context:)
+    if log_level_enabled?(severity)
+      add(severity, message: message, progname: @progname, extra_context: extra_context)
     end
-    progname ||= @progname
-    if message.nil?
-      if block_given?
-        message = yield
-      else
-        message = progname
-        progname = @progname
-      end
-    end
-    write_entry_to_log(severity, Time.now, progname, message, current_context_for_thread.deep_merge(extra_context || {}))
     true
   end
 
+  def add(severity, message:, progname:, extra_context:)
+    write_entry_to_log(severity, Time.now, progname, message, current_context_for_thread.deep_merge(extra_context))
+  end
+
   def write_entry_to_log(severity, timestamp, progname, message, context)
-    @logdev.write(format_message(format_severity(severity), timestamp, progname, message, context))
+    @logdev&.write(format_message(format_severity(severity), timestamp, progname, message, context))
   end
 
   private
 
+  def format_message(severity, timestamp, progname, message, context)
+    message_with_context = message_with_context(context, message, severity, timestamp, progname)
+
+    if @formatter
+      @formatter.call(severity, timestamp, progname, message_with_context)
+    else
+      "#{message_with_context.to_json}\n"
+    end
+  end
+
   def message_with_context(context, message, severity, timestamp, progname)
-    context.merge(
-      message: message,
-      severity: severity,
-      timestamp: timestamp,
-      progname: progname
-    )
+    extra_context =
+      {
+        message: message,
+        severity: severity,
+        timestamp: timestamp
+      }
+    extra_context[:progname] = progname if progname
```

Comment:
I've left `progname` here for now. It sure feels like another name for `log_source` though...

Reply:
why use mutation when we can just compact the hash to remove `progname` when it doesn't exist?

Reply:
This code will be called a _lot_ so I don't think it's good to add things to hash that we didn't want, then use `compact` make a copy with that removed. Also that's too broad, as it could remove other keys if they had `nil` values, which isn't what we intended.
I don't worry too much about mutating local variable hashes to get them to what we want. The local variable is declared just 6 lines up. It's unfortunate that Ruby has no notation to conditionally add keys to hashes, so you _have_ to do this. I could imagine a conditional notation like:
```
    extra_context =
      {
        message: message,
        severity: severity,
        timestamp: timestamp,
        (progname: progname if progname)
      }
```
Oh well.